### ISSUE TITLE
Ignore stable availability during canary updates

### ISSUE THREAD

**Issue by PedroArvela** [argoproj/argo-rollouts#995]
## Summary 

If an existing rollout's revision has containers not ready or in CrashLoopBackoff, attempting to roll out a new version using the Canary strategy will be stuck in "more replicas need to be updated". After several minutes it returns Degraded with "ReplicaSet "_canary revision_" has timed out progressing.", even though all pods of that ReplicaSet are ready.

Steps:

1. Create a Rollout with strategy Canary and health probe and liveness probe both failing.
2. Create a new revision of said rollout with the health probe and liveness probe both working now.
3. Wait for the rollout to progress

### Expected results

Rollout should succeed as the canary revision works and the stable revision does not.

### Actual results

Rollout becomes stuck in step 0/2 for several minutes. After while Rollout becomes degraded and complains that _ReplicaSet "mydeployment-canaryrevision" has timed out progressing_ when in reality it is the stable revision that is failing.

### Additional information

The Rollout we used has a custom image with a purposely broken health and liveness endpoint that always return 500. The "fixed" version returns 200.

The strategy used for the PoC was:

```yaml
  strategy:
    canary:
      steps:
      - setWeight: 50
      - pause:
          duration: 1m
```

Resulting in 2 steps that should work fine in theory.

## Diagnostics

What version of Argo Rollouts are you running?

```
# Paste the logs from the rollout controller

# Logs for the entire controller:
kubectl logs -n argo-rollouts deployment/argo-rollouts
```

https://gist.github.com/PedroArvela/4ce7c449f4e56060d6a3bf2178589be4

```
# Logs for a specific rollout:
kubectl logs -n argo-rollouts deployment/argo-rollouts | grep rollout=borker
```

https://gist.github.com/PedroArvela/face0aa024ff0ee2b56b150c8e6ebeee

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by jessesuen** [comment]
We actually have a test for this (https://github.com/argoproj/argo-rollouts/blob/1c93ebc75131adde85c1086e4a5458e6f666372c/test/e2e/functional_test.go#L371), so I'm surprised you are experiencing this.

Which version of Rollouts?

**Comment by PedroArvela** [comment]
We are using argo rollouts 0.10.2, installed through the manifests in https://github.com/argoproj/argo-rollouts/blob/v0.10.2/manifests/install.yaml

We are using it in an EKS 1.16.15 cluster.

I did some additional tests, and this is what I could find.

I created the docker image specified in this repo: https://github.com/PedroArvela/http-echo/tree/http-fail-endpoint

This is a fork from hashicorp's http-echo image.

That image has a special endpoint `/bork`, which starts returning 500 after a programmable amount of time after the pod starts running (configurable with the `-bork <seconds>` flag).

Deploying [this manifest](https://gist.github.com/PedroArvela/2dec1fdf8a40d6f8f7fc1891da1c3a4b) (with the image set to a local docker repository) quickly deploys and marks the Rollout as Healthy until the pods start restarting. When the pods no longer report being ready and go into CrashLoopBackoff, instead of the Rollout going Degraded, as expected, it goes instead into Progressing. From Progressing it does not appear to go to Degraded.

```
Name:            borker
Namespace:       development
Status:          ◌ Progressing
Message:         updated replicas are still becoming available
Strategy:        Canary
  Step:          2/2
  SetWeight:     100
  ActualWeight:  100
Images:          http-echo-test-2 (stable)
Replicas:
  Desired:       3
  Current:       3
  Updated:       3
  Ready:         0
  Available:     0

NAME                               KIND        STATUS              AGE    INFO
⟳ borker                           Rollout     ◌ Progressing       4m23s  
└──# revision:1                                                           
   └──⧉ borker-bdc554787           ReplicaSet  ◌ Progressing       4m23s  stable
      ├──□ borker-bdc554787-b22hr  Pod         ✖ CrashLoopBackOff  4m23s  ready:0/1,restarts:5
      ├──□ borker-bdc554787-jpvdl  Pod         ✖ CrashLoopBackOff  4m23s  ready:0/1,restarts:5
      └──□ borker-bdc554787-sn2f8  Pod         ✖ CrashLoopBackOff  4m23s  ready:0/1,restarts:5

```

From that progressing state, unlike a Degraded state where the Rollout of a new revision would work as expected, rolling out a revision does not work as the controller continues assuming the previous revision was healthy when it was not. As a result, the rollout of the new canary does not progress unless there is a lucky moment where the stable pods that are in crashloop manage to report as ready all at the same time.

```
Name:            borker
Namespace:       development
Status:          ◌ Progressing
Message:         more replicas need to be updated
Strategy:        Canary
  Step:          0/2
  SetWeight:     50
  ActualWeight:  100
Images:          http-echo-test-2 (canary, stable)
Replicas:
  Desired:       3
  Current:       4
  Updated:       2
  Ready:         2
  Available:     2

NAME                               KIND        STATUS              AGE   INFO
⟳ borker                           Rollout     ◌ Progressing       7m1s  
├──# revision:2                                                          
│  └──⧉ borker-cd46b8c4f           ReplicaSet  ✔ Healthy           54s   canary
│     ├──□ borker-cd46b8c4f-8pvkp  Pod         ✔ Running           54s   ready:1/1
│     └──□ borker-cd46b8c4f-cdncv  Pod         ✔ Running           54s   ready:1/1
└──# revision:1                                                          
   └──⧉ borker-bdc554787           ReplicaSet  ◌ Progressing       7m1s  stable
      ├──□ borker-bdc554787-b22hr  Pod         ✖ CrashLoopBackOff  7m1s  ready:0/1,restarts:6
      └──□ borker-bdc554787-jpvdl  Pod         ✖ CrashLoopBackOff  7m1s  ready:0/1,restarts:6
```

Once again, here are the gists of the logs of the controller during this experiment:

https://gist.github.com/PedroArvela/0bbee47e62f9f73c0ce378a48d726c8f

https://gist.github.com/PedroArvela/9564e66ae3a9c5c5a715d8f92769662b

I hope this brings more insight to the situation. If there's anything I can do to help narrow down this please let me know.

**Comment by zcc35357949** [comment]
Because the stable revision's available replicas can not reach desired replica counts from weight in the first step.
https://github.com/argoproj/argo-rollouts/blob/master/rollout/canary.go#L203
`completedCurrentCanaryStep` always return false,  the rollout will be stuck in progressing state.

https://github.com/argoproj/argo-rollouts/blob/master/test/e2e/functional_test.go#L371
Rollout has no steps in this test, it can not cover this situation.
https://github.com/argoproj/argo-rollouts/blob/master/rollout/canary.go#L302-L303
@jessesuen @PedroArvela

**Comment by zcc35357949** [comment]
Actually, we do not need to check stable revision desired replicas in such case.

**Comment by joberdick-rcd** [comment]
@PedroArvela what's your current work around in this situation? Do you have to delete the entire rollout CRD and start over? Looking for a way to force rollout in this situation that's gitiops friendly.

**Comment by PedroArvela** [comment]
We haven't really encountered this situation naturally after starting to use Rollouts. But our breaking-glass situation that we prepared if we ever find this is to trigger a rollout with no canary steps at all.

That should make the new rollout work more like a classical Kubernetes deployment and replace one version with the other right away without checking if the previous pods are healthy or not.

We tested this a few months ago, so I'm not exactly sure of the details, but if I recall correctly it works and shouldn't be an issue.

**Comment by joberdick-rcd** [comment]
@PedroArvela  Thanks. That's only way I could think of.

**Comment by jessesuen** [comment]
`kubectl argo rollouts promote ROLLOUT --full` is supposed to be the break-glass command to get to the new revision ASAP like Deployments.

**Comment by Sanchit-20** [comment]
> The difficult task is deciding that if "stable" RS is not healthy, and the new replicaset does appear to be healthy, should the Rollout ignore these prescribed canary steps, analysis, etc... and go straight to the new revision? 

When we rollout a new version(say `v4`) and once all the `steps` mentioned in canary(for v4) are completed successfully, so in this case the current default behaviour is the existing rollout gets promoted _**automatically**_  (unless we mention an infinite `pause: {}` in `steps`) to the latest version and version `v4` becomes the stable version.

@jessesuen  Similarly, for the current use case when initial replicaset (`stable`) is unhealthy and canary pods are healthy - Even in this case I think Rollout should get promoted to the new version given all the canary steps are executed successfully. And why do we even want to check if the current ('stable`) version is healthy/not since canary pods are already ready to being promoted to stable?

Also, other issue that I had observed in this `Rollout degradation` scenario is:
One you rollout a new version, once canary pods are up and running, AnalysisRun dont get executed ever (which it should).


> I'd like to understand if manual promote full is sufficient for the use case so we can avoid a feature like:

For our case, I think automatic promotion is very much needed, since we would not want to login into the production environment and run the "break-glass" manually (considering our pipelines and the whole deployment process is completely automated) to skip away from the rollout `degradation` issue.

**Comment by jessesuen** [comment]
This issue is a year old, but revisiting this again since https://github.com/argoproj/argo-rollouts/issues/1774 reminded me again. I reread https://github.com/argoproj/argo-rollouts/issues/995#issuecomment-938029382 and I think are some valid points here. 

Currently, the behavior is that a Rollout will not progress past a step if cannot achieve the needed canary/stable replica ratio to reach the desired canary weight with all pods being available. This check currently takes into consideration the availability of **_both_** canary and stable.

However, it seems that the proposed behavior is that we should ignore stable availability as long as canary availability is meeting its desired weight, and the rollout can continue to proceed with the steps/analysis. 

I understand the arguments for this and I think it's is worth exploring. However, we'll likely need to investigate a solution for 1.3 since 1.2 is wrapping up.

**Comment by jessesuen** [comment]
Change from bug to enhancement because this was original design but something to revisit.

**Comment by jessesuen** [comment]
Had a discussion about this and agree the current behavior is too surprising. We will change the behavior such that when moving to a new version which is good, the availability of the stable should not prevent the update. The changes which need to be made are:

1. when assessing if a `setWeight` step is completed, we will consult: (a) availability count of canary and (b) the pod count of stable (ignoring availability)
2. however, if we are rolling back, we need to do the reverse. i.e. we count (a) availability of stable and (b) pod count of canary (ignoring availability)

**Comment by MarkSRobinson** [comment]
I just got bit by this bug today, so I'm very excited to see a release with this fix.

**Comment by jessesuen** [comment]
The fix is in master and will be part of v1.3.0
